level.py

```diff
@@ -1,44 +1,79 @@
 import ast
+from typing import Optional, List, Dict
+
+class Level(object):
+    NULL_EVENT = (-1, None)
 
-class Level():
     def __init__(self, filename: str=''):
-        self._events = {}
+        self._current_event = 0
+        self._events = []
 
         self._level_name = ''
         self._num_floors = 0
-        self._total_time_sec = 0.
+        self._capacity = 0
+        self._level_finish_move_number = 0
 
         if filename != '':
             self.parse_from_file(filename)
 
-    # todo implement interface for getting
+    def get_num_floors(self) -> int:
+        return self._num_floors
+
+    def get_capacity(self) -> int:
+        return self._capacity
+
+    def get_next_event(self, move: int) -> Optional[Dict[int, List[int]]]:  #TODO type hinting for null event?
+        if move >= self._peek_next_event()[0]:
+            event = self._pop_next_event()
+        else:
+            event = self.NULL_EVENT
+        return event
 
-    # get_pending_event(time: float) -> Dict[int, List[int]]
+    def get_level_finish_move_number(self) -> int:
+        return self._level_finish_move_number
 
-    # get_total_time() -> float
+    def get_level_name(self) -> str:
+        return self._level_name
 
-    # get_level_name() -> str
+    def _event_get(self, i): #TODO is this suss passing back NULL_EVENT?
+        if -len(self._events) <= i < len(self._events):
+            return self._events[i]
+        else:
+            return self.NULL_EVENT
+
+    def _peek_next_event(self):
+        return self._event_get(self._current_event)
+
+    def _pop_next_event(self):
+        event = self._event_get(self._current_event)
+        self._current_event += 1
```

Comment:
I think all this popping/peeking logic is unnecessarily complicated, I think we should rename `get_next_event` to `get_event_for_move` and have behavior that it looks up an event for that move and returns either an event/None, multiple times the same on multiple calls. I know there was discussion of only returning an event once, but I think that only suits the model where events are time/tick-triggered, not move-triggered. I believe the model only calls this class once anyway as is

Reply:
I can't think of how to implement something like adding a passenger in the model class if this returns the same event many times. Do you just trust that the model keeps track of  when it's processed the event? There is almost going to be a next event.

I agree that the popping/peeking can probably be simplified either way though

Reply:
The class calling Level already needs to know what move we're on. To implement a remove of `move` as a parameter from`get_next_event` we'd have to have some guarantee about it only being called once per move or something anyway

Reply:
The state-transition loop is a bit messy, we can definitely extract move-counting out of it at some point, but it can't be in Level

Reply:
I don't think it's possible to remove the `move` parameter and still have the code make sense, I just realllly don't like a bool in `Bellhop` that would just state "have I acted on this move yet" even though most moves would return a None in the first call

Reply:
I think eventually Bellhop should have a concept where there is a set of logic executed exactly once upon transition to a new state, which I believe currently exists despite being a bit hidden, then it's easy

Reply:
> I think eventually Bellhop should have a concept where there is a set of logic executed exactly once upon transition to a new state, which I believe currently exists despite being a bit hidden, then it's easy

The state machine transitions aren't "moves" though, a move would be getting to a new floor which is why it's in ARRIVING now

Reply:
Never mind, I get it now. I missed the part that the model asks exactly once for each move number